### ISSUE TITLE
Resume "download template" button under images/Templates/zones

### ISSUE THREAD

**Issue by gg4126** [apache/cloudstack#6647]
##### ISSUE TYPE

 * Improvement Request
 * Enhancement Request

##### COMPONENT NAME

~~~
UI
~~~

##### CLOUDSTACK VERSION

~~~
4.17.0.1
~~~

##### CONFIGURATION

Advanced Networking

##### OS / ENVIRONMENT

RHEL-7.9

##### SUMMARY

While a template is being uploaded/pushed to a specific zone due to network issues the image/template was not fully downloaded onto that zone. Currently the download percentage is showing as 31%, and even though the management server was bounced yet the download is not getting completed and am looking for a button to resume the download but there is none.

##### STEPS TO REPRODUCE

~~~
step1: Navigate to images/templates and click "Register template from URL"
step2: Fill-up the form to download the template to the desired zone and click OK
step3: While the template is being downloaded - due to network outages or the apache server where qcow2 templates are kept suddenly goes down this will interrupt the template downloading to a specific zone(s).
step4: Once the network outage is resolved, the user should be in a position to "resume download" on images/Templates/zones tab which is NOT available at the moment, hence creating this enhancement request
~~~

##### EXPECTED RESULTS

~~~
As soon as the network issue is resolved, the user should be able to resume the template download by clicking "Resume Download" button. 
~~~

##### ACTUAL RESULTS

~~~
there is no button to resume the template download; 
~~~
![image](https://user-images.githubusercontent.com/61292595/184906745-8d6e99f5-cf89-4450-966d-b8a1ad1252f8.png)

**Comment by gg4126** [comment]
[
![image](https://user-images.githubusercontent.com/61292595/184962684-887e50d0-e5d4-49e5-87ee-b9893b19825c.png)
](url)

**Comment by gg4126** [comment]
[](url
![template-download-issues](https://user-images.githubusercontent.com/61292595/184962789-1f2b4f2d-f805-4d47-9665-b9509759a454.png)
)

**Comment by rohityadavcloud** [comment]
I think if you're uploading an ISO or template and something goes wrong with SSVM and mgmt server, it might loose track of it and you'll need to redo the upload (there's a timeout after which the upload will be marked as fail).

**Comment by gg4126** [comment]
Thanks Rohit for your response! the alternate solution works, and we are good with it for now, but please consider having this button to resume the download if it is possible.

**Comment by gitsan13** [comment]
I would like to contribute to this enhancement. Can you guide me through the process?

**Comment by gg4126** [comment]
@gitsan13,
Thanks for your interest on this enhancement, please get in touch with @rohityadavcloud - he can guide you thru...thanks!

**Comment by daviftorres** [comment]
I have reports from clients who tried to download a Template or ISO from a URL, but the download either hangs indefinitely or takes a very long time.

Because I have limited Java skills, I couldn’t confidently find the exact part of the code where the download happens. However, I found one section that I can use to explain my proposal:

<img width="1556" height="505" alt="Image" src="https://github.com/user-attachments/assets/2fcf785c-d6cb-419a-881a-ff15974dcb65" />

In this example, I try to prevent the download from hanging by adding a timeout when there’s no data stream for a certain period. I also added a retry mechanism with up to 3 attempts to handle the issue more gracefully.

In short, this is a problem I’ve seen happen multiple times.

**Comment by davift** [comment]
@DaanHoogland ,

Just to clarify the solution that I proposed above: I did not test this because my Java skills are close to none.

I wrote it down based on the symptoms I’ve noticed:

- The download sometimes freezes and doesn’t recover quickly.
- It also doesn’t seem to retry if it fails.

@nvazquez is the best person to assess this, as it relates to his talk about "Bypass CloudStack Secondary Storage."

**Comment by DaanHoogland** [comment]
> [@nvazquez](https://github.com/nvazquez) is the best person to assess this

He is certainly one of them ;) @shwstppr @harikrishna-patnala might have opinions as well ...

**Comment by shwstppr** [comment]
IMO, I see multiple issues here:
- The original improvement request may need introduction of a new API - reconcile/resumeTemplate, etc
- Issue and the solution shared in https://github.com/apache/cloudstack/issues/6647#issuecomment-3140781750 by @daviftorres looks fine a from high-level,  though we might already have some configs for timeout values used. Also, we would need reproduction steps to implement and test this.

cc @DaanHoogland @davift

**Comment by nvazquez** [comment]
Hi @davift @shwstppr - I think adding a retrying mechanism would be helpful but perhaps wouldn't prevent the issue entirely. I agree we could add a new API to manually trigger the resume (or maybe re-download) of a specific template when the issue persists.

@davift were you able to reproduce the issue manually to replicate the symptoms from the step 3? It would be useful for exploring a solution

**Comment by daviftorres** [comment]
Yes, I do have one client who reported today the same issue described in "step 3" by the original author (who suspected a network failure or an issue on the HTTP server).

<img width="1493" height="270" alt="Image" src="https://github.com/user-attachments/assets/395c7aa8-3182-46b7-9a5d-9aea0fc00fcd" />

In this case, I confirmed the problem was indeed on the HTTP server side: the file failed to download from three different sources simultaneously, and I was also unable to download it manually.

On the SSVM, there's no visible activity from a Java application, and the download progress remains frozen at the same value, it doesn't timeout or retry.

<img width="714" height="387" alt="Image" src="https://github.com/user-attachments/assets/440fbd6c-51e6-410c-9d18-ff2ffa22205e" />

To test how the system handles recovery, I re-deployed the SSVM in one of the zones.

As expected, the system quickly detects that the SSVM is not up:

<img width="1058" height="222" alt="Image" src="https://github.com/user-attachments/assets/65e07476-60b0-4650-b52c-682fbc6003ab" />

However, the status provided does not indicate what it's about to do next:

<img width="1022" height="222" alt="Image" src="https://github.com/user-attachments/assets/357c7144-048b-4054-90c7-66741345b76e" />

Finally, it retries the download. As expected, this attempt also fails because the source is still unreachable. Yesterday, I had a similar scenario where the source was available, and the same procedure (re-deploying the SSVM) worked successfully.

<img width="1023" height="219" alt="Image" src="https://github.com/user-attachments/assets/9a2d18df-0fe6-4450-9841-a759268556e7" />

@nvazquez, does this clarify your question regarding "step 3"? I still have the SSVMs in the other two zones running, let me know if you’d like me to collect specific logs that could be helpful.

In summary: the goal here isn’t to fix a bug in the SSVM itself, but to improve its resilience against external failures and ensure it can fail gracefully if self-healing isn’t possible.

**Comment by daviftorres** [comment]
@nvazquez, the only reference to the download that I found in `/var/log/cloud.log` (SSVM) is the following line, which indicates that it has started, but there is nothing after that to show whether it succeeded or failed:

```
2025-08-01T02:26:24,986 INFO  [storage.template.HttpTemplateDownloader] (AgentRequest-Handler-5:[]) No credentials configured for host=<redacted>:80
2025-08-01T02:26:25,088 INFO  [storage.template.HttpTemplateDownloader] (pool-2-thread-4:[]) Starting download from http://<redacted>/images/windows/windows_11_pro-202508010000.qcow2 to /mnt/SecStorage/<redacted>/template/tmpl/4/401/dnld14596858942789915474tmp_ remoteSize=(18.93 GB) 20330577920 , max size=(50.00 GB) 53687091200
```

In comparison, this is how it downloaded an Ubuntu ISO hours later on the same zone with the same SSVM:

```
2025-08-01T13:21:21,194 INFO  [storage.template.HttpTemplateDownloader] (AgentRequest-Handler-7:[]) No credentials configured for host=mirror.0xem.ma:443
2025-08-01T13:21:21,467 INFO  [storage.template.HttpTemplateDownloader] (pool-2-thread-7:[]) Starting download from https://mirror.0xem.ma/ubuntu-releases/24.04.2/ubuntu-24.04.2-desktop-amd64.iso to /mnt/SecStorage/<redacted>/template/tmpl/2/407/dnld12640683685705244346tmp_ remoteSize=(5.91 GB) 6343219200 , max size=(50.00 GB) 53687091200
2025-08-01T13:21:42,232 INFO  [storage.resource.NfsSecondaryStorageResource] (AgentRequest-Handler-6:[]) Determined host 10.4.52.221 corresponds to IP 10.4.52.221
2025-08-01T13:21:45,633 INFO  [storage.resource.NfsSecondaryStorageResource] (AgentRequest-Handler-3:[]) Determined host 10.4.52.221 corresponds to IP 10.4.52.221
2025-08-01T13:22:38,138 INFO  [storage.template.DownloadManagerImpl] (pool-2-thread-7:[]) Download Completion for jobId: c2f8078e-bf54-4a28-90a0-c3be7520663c, status=DOWNLOAD_FINISHED
2025-08-01T13:22:38,146 INFO  [storage.template.DownloadManagerImpl] (pool-2-thread-7:[]) local: /mnt/SecStorage/<redacted>/template/tmpl/2/407/dnld12640683685705244346tmp_, bytes=(5.91 GB) 6343219200, error=Downloaded (5.91 GB) 6343219200 bytes (download complete remote=unknown bytes), pct=100
2025-08-01T13:22:38,146 INFO  [storage.template.DownloadManagerImpl] (pool-2-thread-7:[]) No checksum available for 'dnld12640683685705244346tmp_'
```

**Comment by nvazquez** [comment]
Thanks @daviftorres sorry for the delay, yes it seems like the cases in which the error comes from the server are not properly covered. Will explore a solution based on your input
